binstat/binstat.go

```diff
@@ -0,0 +1,546 @@
+package binstat
+
+/*
+#if defined(__linux__)
+#include <stdint.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+#ifdef SYS_gettid
+uint64_t gettid() { return syscall(SYS_gettid); }
+#else
+#error "SYS_gettid unavailable on this system"
+#endif
+#elif __APPLE__ // http://elliotth.blogspot.com/2012/04/gettid-on-mac-os.html
+#include <stdint.h>
+#include <pthread.h>
+uint64_t gettid() { uint64_t tid; pthread_threadid_np(NULL, &tid); return tid; }
+#else
+#   error "Unknown platform; __linux__ or __APPLE__ supported"
+#endif
+*/
+import "C"
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"runtime/debug"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/rs/zerolog"
+)
+
+// todo: consider using something faster than runtime.nano() e.g. [1]
+// But is [1] maybe not safe with multiple CPUs? [2]
+// "Another concern is that if a thread is migrated on a different
+//  processor between two measurements, the counter might skip too much
+//  or even 'go back'." [3]
+// [1] https://github.com/templexxx/tsc/issues/8
+// [2] https://stackoverflow.com/questions/3388134/rdtsc-accuracy-across-cpu-cores
+// [3] https://coherent-labs.com/posts/timestamps-for-performance-measurements/
+
+type globalStruct struct {
+	dumps            uint64
+	dmpName          string
+	dmpPath          string
+	cutPath          string
+	processBaseName  string
+	processPid       int
+	index            int
+	indexInternalMax int
+	key2index        map[string]int
+	keysArray        []string
+	keysEgLoc        []string
+	frequency        []uint64
+	frequencyShadow  []uint64
+	accumMono        []uint64
+	accumMonoShadow  []uint64
+	verbose          bool
+	enable           bool
+	second           uint64
+	startTime        time.Time
+	startTimeMono    time.Duration
+	lenWhat          string         // e.g. "~Code=99;~X=99"
+	what2len         map[string]int // e.g. Code -> 99, X -> 99
+}
+
+type BinStat struct {
+	what           string
+	enterTime      time.Duration
+	callerFunc     string
+	callerLine     int
+	callerParams   string
+	callerTime     bool
+	callerSize     int64
+	callerSizeWhen int
+}
+
+const (
+	sizeAtEnter = iota
+	sizeNotUsed
+	sizeAtLeave
+)
+
+const (
+	internalSec = iota
+	internalEnter
+	internalPoint
+	internalx_2_y
+	internalDebug
+	internalGC
+)
+
+var globalRWMutex = sync.RWMutex{}
+var global = globalStruct{}
+var globalLog zerolog.Logger
```

Comment:
can't these be part of the global struct and not global parameters ?

Reply:
@ramtinms Thanks for the comment! I agree with you and I will try to refactor this. The mutex was originally outside because I later learned about `init()` and until then I needed the mutex outside of the struct initialization. However, I neglected to refactor the RW lock into the struct after I converted to using `init()` instead. The logger was added later too. But not sure when I didn't put it into the struct. Stay tuned for the update... :-)

Reply:
@ramtinms FYI just folded them into the global struct :-)

Reply:
thanks